core/src/main/java/org/infinispan/transaction/TransactionTable.java

```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
     * on a remote node. This might cause leaks, e.g. if the transaction is prepared, committed and prepared again.
     * Once marked as completed (because of commit or rollback) any further prepare received on that transaction are discarded.
     */
-   public void markTransactionCompleted(GlobalTransaction globalTx) {
-      completedTransactions.put(globalTx, System.nanoTime());
-   }
+    public void markTransactionCompleted(GlobalTransaction globalTx) {
+        if (completedTransactionsInfo != null) {
```

Comment:
after it is started, the `completedTransactionsInfo` can't be null - why the check then?